File: web/src/lib/session/security-notices.ts
// The in-memory hand-off for the login security notice (auth.md "Reuse
// detection"): login navigates away on the phase flip, so the signed-in
// shell renders the notice from this per-tab store rather than the
// login surface. A tab closed before rendering loses it — the accepted
// narrow-carrier trade-off (auth.md).

export type SecurityNotices = {
  /** The pending detection time (ISO 8601); null once dismissed or when none arrived. */
  reuseDetectedAt(): string | null;
  post(detectedAt: string): void;
  dismiss(): void;
  subscribe(listener: () => void): () => void;
};

export function createSecurityNotices(): SecurityNotices {
  let reuseDetectedAt: string | null = null;
  const listeners = new Set<() => void>();

  const notify = () => listeners.forEach((listener) => listener());

  return {
    reuseDetectedAt: () => reuseDetectedAt,
    post(detectedAt) {
      reuseDetectedAt = detectedAt;
      notify();
    },
    dismiss() {
      reuseDetectedAt = null;
      notify();
    },
    subscribe(listener) {
      listeners.add(listener);
      return () => listeners.delete(listener);
    },
  };
}

export const securityNotices: SecurityNotices = createSecurityNotices();

File: android/core/domain/src/main/kotlin/com/cogra/domain/identity/SecurityNotices.kt
// The in-memory hand-off for the login security notice (auth.md "Reuse
// detection"): the backend delivers the event exactly once on the
// LogInPayload, and login navigates away immediately, so the app shell
// renders it from this holder rather than the login surface. Process
// death before rendering loses the notice — the accepted narrow-carrier
// trade-off (auth.md).

package com.cogra.domain.identity

import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

@Singleton
class SecurityNotices @Inject constructor() {

    private val _reuseDetectedAt = MutableStateFlow<Instant?>(null)

    /** The pending notice; null once dismissed or when none arrived. */
    val reuseDetectedAt: StateFlow<Instant?> = _reuseDetectedAt.asStateFlow()

    fun post(detectedAt: Instant) {
        _reuseDetectedAt.value = detectedAt
    }

    fun dismiss() {
        _reuseDetectedAt.value = null
    }
}
